### PR TITLE
Fix SimpleExemplarReservoir for cumulative

### PR DESCRIPTION
**Pull request by cijothomas**
Fixes `SimplerExemplarReservoir` for cumulative temporality to reset internal state during collect. Before this change, newer measurements had no fair chance to make it to the reservoir. This PR resets the "number_of_measurements_seen" with every collect, so new measurements has fair chance to make it to reservoir, subject to the reservoir sampling algorithm.

Spec ref: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#simplefixedsizeexemplarreservoir

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs [hunk 1]
```diff
@@ -57,8 +57,12 @@ public override Exemplar[] Collect(ReadOnlyTagCollection actualTags, bool reset)
             if (reset)
             {
                 this.runningExemplars[i].Timestamp = default;
-                this.measurementsSeen = 0;
             }
+
+            // Reset internal state irrespective of temporality.
+            // This ensures incoming measurements have fair chance
+            // of making it to the reservoir.
+            this.measurementsSeen = 0;
```
Comment (rajkumar-rangaraj) [inline src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs:65]: Should this be reset on every iteration of the for loop?

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs]: good catch! fixed.

File: src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs [hunk 2]
```diff
@@ -57,9 +57,13 @@ public override Exemplar[] Collect(ReadOnlyTagCollection actualTags, bool reset)
             if (reset)
```
Comment (alanwest) [inline src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs:57]: unrelated but kind of related nit: maybe this variable should be named `isDelta` for readability?

**Review by alanwest — APPROVED** [review]

File: test/OpenTelemetry.Tests/Metrics/MetricExemplarTests.cs [hunk 3]
```diff
@@ -48,6 +50,12 @@ public void TestExemplarsCounter()
         Assert.True(metricPoint.Value.StartTime >= testStartTime);
         Assert.True(metricPoint.Value.EndTime != default);
         var exemplars = GetExemplars(metricPoint.Value);
+
+        // TODO: Modify the test to better test cumulative.
+        // In cumulative where SimpleExemplarReservoir's size is
+        // more than the count of new measurements, it is possible
+        // that the exemplar value is for a measurement that was recorded in the prior
+        // cycle. The current ValidateExemplars() does not handle this case.
```
Comment (alanwest) [inline test/OpenTelemetry.Tests/Metrics/MetricExemplarTests.cs:58]: Removing trailing space...

```suggestion
        // cycle. The current ValidateExemplars() does not handle this case.
```

File: src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs [hunk 4]
```diff
@@ -57,9 +57,13 @@ public override Exemplar[] Collect(ReadOnlyTagCollection actualTags, bool reset)
...
                 this.runningExemplars[i].Timestamp = default;
-                this.measurementsSeen = 0;
             }
-        }
+        }
+
+        // Reset internal state irrespective of temporality.
+        // This ensures incoming measurements have fair chance
+        // of making it to the reservoir.
+        this.measurementsSeen = 0;
```
Comment (alanwest) [inline src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs:66]: Trying to fix line endings via github too 🤞...

```suggestion
        }

        // Reset internal state irrespective of temporality.
        // This ensures incoming measurements have fair chance
        // of making it to the reservoir.
        this.measurementsSeen = 0;
```

Comment (alanwest) [reply src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs:66]: hehe. didn't work. @cijothomas fix your line endings 😄

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs:66]: oops. I used a temporary cloud box to code this, and that machine has something weird about line endings! Let me fix it,

**Review by rajkumar-rangaraj — APPROVED** [review]

File: src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs (hunk 2 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs:57]: yes... The Reservoir itself is not necessarily aware of temporality concept. All it needs to know is if it should reset its internal state or not.. The caller knows temporality and passes reset to this...

Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs:57]: Not a blocker for this PR, but just looking at this API shape, is the correct thing to do to expose `bool reset`?

> SDKs are free to decide whether "collect" should also reset internal storage for delta temporal aggregation collection, or use a more optimal implementation.

Curious what other languages have done to solve this.

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Exemplar/SimpleExemplarReservoir.cs:57]: The current APIs need to be revised before exposing them to public (its marked internal), and there are many TODOs to be solved.

**Pull request merged**